include/watchdog.h

```diff
@@ -1,116 +1,219 @@
  */
-struct wdt_config {
-	u32_t timeout;
-	enum wdt_mode mode;
-	void (*interrupt_fn)(struct device *dev);
+struct wdt_window {
+	u32_t min;
+	u32_t max;
 };
 
-typedef void (*wdt_api_enable)(struct device *dev);
-typedef void (*wdt_api_disable)(struct device *dev);
-typedef int (*wdt_api_set_config)(struct device *dev,
-				  struct wdt_config *config);
-typedef void (*wdt_api_get_config)(struct device *dev,
-				   struct wdt_config *config);
-typedef void (*wdt_api_reload)(struct device *dev);
+/** Watchdog callback. */
+typedef void (*wdt_callback_t)(struct device *dev, void *handle);
 
-struct wdt_driver_api {
-	wdt_api_enable enable;
-	wdt_api_disable disable;
-	wdt_api_get_config get_config;
-	wdt_api_set_config set_config;
-	wdt_api_reload reload;
+/**
+ * @brief Watchdog timeout configuration struct.
+ *
+ * @param window Timing parameters of watchdog timeout.
+ * @param handle Unique handle used to reload specific watchdog timeout.
+ *		 This handle is filled by the user, it will be supplied to
+ *		 callback if WDT_FLAG_RUN_CALLBACK is set. Good example is
+ *		 handle pointing to the id of the thread that is resposible for
+ *		 reloading corresponding timeout.
+ * @param callback Timeout callback.
+ * @param next Pointer to the next timeout configuration. It is used for
+ *	       watchdogs with staged timeouts functionality. Should be NULL
+ *	       for single stage timeout.
+ * @param flags Bit field with following parts:
+ *
+ *     callback     [ 0 ]       - run user callback after timeout.
+ *     reset        [ 1 : 2 ]   - perform specified reset after timeout/callback.
+ */
+struct wdt_timeout_cfg {
+	wdt_window window;
+	void *handle;
+	wdt_callback_t callback;
+	struct wdt_timeout_cfg *next;
+	u8_t flags;
 };
 
-static inline void wdt_enable(struct device *dev)
-{
-	const struct wdt_driver_api *api = dev->driver_api;
-
-	api->enable(dev);
-}
+/**
+ * @brief Watchdog driver API.
+ */
+struct wdt_driver_api {
+	int (*set_up)(struct device *dev, u8_t options);
+	int (*disable)(struct device *dev);
+	int (*install_timeout)(struct device *dev, struct wdt_timeout_cfg *cfg);
+	int (*reload)(struct device *dev, void *handle);
+};
 
-static inline void wdt_disable(struct device *dev)
+/**
+ * @brief Set up watchdog instance.
+ *
+ * This function is used for configuring global watchdog settings that
+ * affect all timeouts. It should be called after installing timeouts.
+ * After successful return all installed timeouts are valid.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param options Bit field with following parts:
+ *
+ *     pause in sleep      [ 0 ]       - pause in CPU sleep
+ *     pause by debugger   [ 1 ]       - pause when halted by debugger
+ *
+ * @retval 0 If success.
+ * @retval -ENOTSUP If any of the set options is not supported.
+ * @retval -EBUSY If watchdog instance has been already setup.
+ */
+static inline int wdt_set_up(struct device *dev, u8_t options)
```

Comment:
From a grammar point of view, "setup" is a noun (or adjective) while "set up" is a verb (phrase), so you can "set up a chess board", or you can "go to a setup screen",  or you can "pay a setup fee"  (generally not used with a hyphen.)  Alas, @mnkp is probably right with the convention of using "setup" rather than "set_up" in function names even if it's an action.